crates/iota-json-rpc/src/governance_api.rs

```diff
@@ -636,10 +648,108 @@ fn pending_validator_exchange_rate(
+/// changes only at epoch change.
+fn candidate_validators_exchange_rate(
+    state: &Arc<dyn StateRead>,
+) -> RpcInterimResult<Vec<ValidatorExchangeRates>> {
+    let system_state_summary = state.get_system_state()?.into_iota_system_state_summary();
+
+    // From validator_candidates_id table get validator info using as key its
+    // IotaAddress
+    let tables = validator_summary_from_system_state(
+        state,
+        system_state_summary.validator_candidates_id,
+        system_state_summary.validator_candidates_size,
+        |df| bcs::from_bytes::<IotaAddress>(&df.bcs_name).map_err(Into::into),
+    )?;
+
+    validator_exchange_rates(state, tables)
+}
+
+/// Fetches validator status information from `StateRead` for validators not
+/// included in `IotaSystemStateSummary`.
+///
+/// `IotaSystemStateSummary` only contains information about `Active`
+/// validators. To retrieve information about `Inactive`, `Candidate`, and
+/// `Pending` validators, we need to access dynamic fields within specific
+/// tables.
+///
+/// To retrieve validator status information, this function utilizes a
+/// `table_id` defined by an `ObjectID` and a `limit` to specify the number of
+/// records to fetch. Both the `table_id` and `limit` are obtained from
+/// `IotaSystemStateSummary`. Additionally, a `key` is extracted from the
+/// `table_id`'s `DynamicFieldInfo` to identify the specific validator within
+/// the table.
+///
+/// # Example
+///
+/// ```rust
+/// // Get inactive validators
+/// let system_state_summary = state.get_system_state()?.into_iota_system_state_summary();
+/// let _ = validator_summary_from_system_state(
+///     state,
+///     // ID of the object that maps from a staking pool ID to the inactive validator that has that pool as its staking pool.
+///     system_state_summary.inactive_pools_id,
+///     // Number of inactive staking pools.
+///     system_state_summary.inactive_pools_size,
+///     // Extract the `ID` of the `Inactive` validator from the `DynamicFieldInfo` in the `system_state_summary.inactive_pools_id` table
+///     |df| bcs::from_bytes::<ID>(&df.bcs_name).map_err(Into::into),
+/// ).unwrap();
+/// ```
+///
+/// # Example
+///
+/// ```rust
+/// // Get candidate validators
+/// let system_state_summary = state.get_system_state()?.into_iota_system_state_summary();
+/// let _ = validator_summary_from_system_state(
+///     state,
+///     // ID of the object that stores preactive validators, mapping their addresses to their Validator structs
+///     system_state_summary.validator_candidates_id,
+///     // Number of preactive validators
+///     system_state_summary.validator_candidates_size,
+///     // Extract the `IotaAddress` of the `Candidate` validator from the `DynamicFieldInfo` in the `system_state_summary.validator_candidates_id` table
+///     |df| bcs::from_bytes::<IotaAddress>(&df.bcs_name).map_err(Into::into),
+/// ).unwrap();
+/// ```
```

Comment:
Much much better, thank you!

Reply:
Thank to you @kodemartin, really appreciated the suggestions, will keep in mind next time when something similar will occur 🙏